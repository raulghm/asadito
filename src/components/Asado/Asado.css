.Asado {
  max-width: 800px;
  margin: 10px auto;
}

.Asado-wrapper {
	text-align: center;
	padding: 20px;
	color: #fff;
  position: relative;
  z-index: 1;
}

.Asado-wrapper:before {
  content: '';
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-image: url(../../assets/img/img-chalkboard.jpg);
  background-size: cover;
  background-position: 50%;
  opacity: .9;
  box-shadow: 0 0 9px 1px rgba(0,0,0,.6);
}

@media (min-width: 30em) {
  .Asado-wrapper:before {
    border-radius: 10px;
  }
}

@media (min-width: 39.8em) {
  .Asado-wrapper {
    padding: 40px 60px;
  }
}

.Asado-body {
  position: relative;
}

.Asado h1 {
	font-family: 'Abril Fatface';
	font-weight: lighter;
	font-size: 2rem;
	letter-spacing: 2px;
	margin: 20px 0;
}

@media (min-width: 39.8em) {
  .Asado h1 {
    font-size: 3rem;
  }
}

.Asado h2 {
	font-size: 1.6rem;
}

@media (min-width: 39.8em) {
  .Asado h2 {
    font-size: 2rem;
  }
}

.Asado h3 {
  font-size: 1.2rem;
  font-weight: bold;
}

@media (min-width: 39.8em) {
  .Asado h3 {
    font-size: 1.4rem;
  }
}

.Asado h4 {
	margin: 10px;
}

.Asado input[type="number"] {
  width: 90%;
  max-width: 60px;
}

.Asado input[type="number"],
.Asado select {
	font-size: 1.4rem;
	border-radius: 4px;
	border: 0;
	padding: .2em;
	outline: 0;
}

.Asado-select {
	padding-bottom: 20px;
	border-bottom: 1px solid rgba(255,255,255,.2);
}

.Asado-select:last-child {
	border: 0;
}

.Asado-input {
  margin-bottom: 14px;
  user-select: none;
}

.Asado-input--checkbox {
  margin-bottom: 10px;
}

.Asado-input--checkbox label {
  cursor: pointer;
}

.Asado-input--checkbox input[type="checkbox"] {
  width: 16px;
  height: 16px;
  margin-left: 10px;
  display: inline-block;
}

.Asado-select-item h4 span {
  display: block;
}

@media (min-width: 39.8em) {
  .Asado-select-item h4 span {
    display: inline-block;
  }
}

.Asado-select-body--column {
	display: grid;
	grid-auto-flow: column;
}

.Asado hr {
	border: 0;
	border-bottom: 1px solid #fff;
	opacity: .4;
}

.Asado li {
  list-style: none;
}

.Asado span[role="img"] {
  margin-left: 4px;
}
