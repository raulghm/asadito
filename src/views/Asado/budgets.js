/**
 * Valores extraidos desde la web de supermercados Jumbo y Lider
 * promediando ambos valores.
 * Solo carnes al vacio y con precios normales.
 * Se segmentan presupuestos en base a sus valores.
 */
const budgets = [
  {
    id: 0,
    name: 'Económico',
    meats: [
      {
        name: 'Abastero',
        value: 4490
      },
      {
        name: 'Carnicero',
        value: 4490
      },
      {
        name: 'Tapapecho',
        value: 4790
      },
      {
        name: 'Tapabarriga',
        value: 5740
      },
      {
        name: 'Asado de tira',
        value: 5490
      },
      {
        name: 'Sobre costilla',
        value: 5190
      },
      {
        name: 'Huachalomo',
        value: 4490
      },
      {
        name: 'Punta paleta',
        value: 4640
      }
    ],
    sausage: 'Chorizos o salchichas',
    sausagePrice: 3200
  },
  {
    id: 1,
    name: 'Medio',
    meats: [
      {
        name: 'Plateada',
        value: 5340
      },
      {
        name: 'Punta picana',
        value: 5990
      },
      {
        name: 'Asiento',
        value: 7140
      },
      {
        name: 'Lomo liso',
        value: 6490
      },
      {
        name: 'Punta de ganso',
        value: 8740
      }
    ],
    sausage: 'Chorizos',
    sausagePrice: 4000
  },
  {
    id: 2,
    name: 'Alto',
    meats: [
      {
        name: 'Lomo vetado',
        value: 9190
      },
      {
        name: 'Palanca',
        value: 10490
      },
      {
        name: 'Filete',
        value: 10240
      },
      {
        name: 'Entraña',
        value: 13490
      }
    ],
    sausage: 'Longanizas',
    sausagePrice: 5800
  },
  {
    id: 3,
    name: 'Premium',
    meats: [
      {
        name: 'Wagyu',
        value: 15990
      },
      {
        name: 'Angus',
        value: 24990
      },
      {
        name: 'Filete premium',
        value: 15990
      },
      {
        name: 'Lomo vetado premium',
        value: 14990
      },
      {
        name: 'Otros cortes premium',
        value: 14990
      }
    ],
    sausage: 'Longanizas premium',
    sausagePrice: 9000
  }
]

export default budgets
