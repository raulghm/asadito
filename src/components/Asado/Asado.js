import React, { Component } from 'react'
import Footer from '../Footer/Footer'
import './Asado.css'

class Asado extends Component {
  constructor(props) {
    super(props)

    this.state = {
      title: '¬øCu√°nto debo comprar para el asado?',
      user: {
        men: 0,
        women: 0,
        children: 0,
      },
      breadPrice: 1200,
      carbonPrice: 2800,
      budgets: [
        {
          id: 0,
          name: 'Econ√≥mico',
          options: [
            'Abastero',
            'Carnicero',
            'Tapapecho',
            'Tapabarriga',
            'Asado de tira',
            'Sobre costilla',
            'Huachalomo',
            'Punta paleta'
          ],
          sausage: 'Chorizos o salchichas',
          meatPrice: 5000,
          sausagePrice: 3200
        },
        {
          id: 1,
          name: 'Medio',
          options: [
            'Plateada',
            'Punta picana',
            'Asiento',
            'Lomo liso',
            'Punta de ganso'
          ],
          sausage: 'Chorizos',
          meatPrice: 7200,
          sausagePrice: 4000
        },
        {
          id: 2,
          name: 'Alto',
          options: [
            'Lomo vetado',
            'Palanca',
            'Filete',
            'Entra√±a'
          ],
          sausage: 'Longanizas',
          meatPrice: 10670,
          sausagePrice: 5800
        },
        {
          id: 3,
          name: 'Premium',
          options: [
            'Wagyu',
            'Angus',
            'Lomo vetado premium'
          ],
          sausage: 'Longanizas',
          meatPrice: 25000,
          sausagePrice: 7000
        }
      ],
      budgetIndex: 0,
      budgetSelected: null,
    }
  }

  handleMen = event => this.setState({
    user: {...this.state.user, men: parseInt(event.target.value, 10) }
  })
  handleWomen = event => this.setState({
    user: {...this.state.user, women: parseInt(event.target.value, 10) }
  })
  handleChildren = event => this.setState({
    user: {...this.state.user, children: parseInt(event.target.value, 10) }
  })

  handleBudget = event => {
    const value = parseInt(event.target.value, 10)
    const budget = this.state.budgets.find(item => item.id === value)

    this.setState({
      budgetIndex: event.target.value,
      budgetSelected: budget
    })
  }

  peopleCount = () => {
    return this.state.user.men +
          this.state.user.women +
          this.state.user.children
  }

  meatCount = () => {
    return this.round(
      (this.state.user.men * 0.35) +
      (this.state.user.women * 0.25) +
      (this.state.user.children * 0.2)
    )
  }

  sausageCount = () => {
    return (this.state.user.men * 0.1) +
      (this.state.user.women * 0.05) +
      (this.state.user.children * 0.05)
  }

  round = (number, precision = 1) => {
    const factor = Math.pow(10, precision)
    const result = Math.round(number * factor) / factor
    return number ? result : 0
  }

  totalPrice = () => {
    const meat = this.meatCount() * this.state.budgetSelected.meatPrice
    // const sausage = this.sausageCount() * this.state.budgetSelected.sausagePrice
    return this.round(meat)
  }

  eachPrice = () => {
    return this.round(
      this.totalPrice() /
      (this.state.user.men + this.state.user.women)
    , -1)
  }

  render() {
    const {
      title,
      budgets,
      budgetSelected,
      peopleCount = this.peopleCount(),
    } = this.state

    return (
      <div className="Asado">
        <div className="Asado-wrapper">
          <div className="Asado-body">
            <h1>Calculadora de Asado <span role="img" aria-label="meat">üçñ</span></h1>

            <h2>{title}</h2>

            <div className="Asado-select">
              <h3>1- Indica la cantidad de comensales</h3>

              <div className="Asado-select-body Asado-select-body--column">
                <div className="Asado-select-item">
                  <h4>Hombres <span role="img" aria-label="men">üë®</span></h4>
                  <input
                    type="number"
                    pattern="[0-9]*"
                    min={0}
                    max={100}
                    placeholder="0"
                    onChange={this.handleMen.bind(this)}
                  />
                </div>

                <div className="Asado-select-item">
                  <h4>Mujeres <span role="img" aria-label="women">üë©</span></h4>
                  <input
                    type="number"
                    pattern="[0-9]*"
                    min={0}
                    max={100}
                    placeholder="0"
                    onChange={this.handleWomen.bind(this)}
                  />
                </div>

                <div className="Asado-select-item">
                  <h4>Ni√±os <span role="img" aria-label="children">üë∂</span></h4>
                  <input
                    type="number"
                    pattern="[0-9]*"
                    min={0}
                    max={100}
                    placeholder="0"
                    onChange={this.handleChildren.bind(this)}
                  />
                </div>
              </div>

              <h3>Total comensales: {peopleCount}</h3>
            </div>

            { peopleCount > 0 &&
              <div className="Asado-select">
                <h3>2- Selecciona tu presupuesto</h3>

                <div className="Asado-select-body Asado-select-body--column">
                  <div className="Asado-select-item">
                    <select onChange={this.handleBudget.bind(this)}>
                      <option>Selecciona...</option>
                      { budgets.map(item => {
                        return <option
                          value={item.id}
                          key={item.id}
                        >{item.name}</option>
                      }) }
                    </select>
                  </div>
                </div>
              </div>
            }

            { peopleCount > 0 && budgetSelected &&
              <div className="Asado-select">
                <h3>3- Resultados</h3>

                <div className="Asado-select-body">
                  Opciones de carnes para comprar:
                  <p>{ budgetSelected.options.map((option, i) => {
                    if (i === budgetSelected.options.length - 1) {
                      return `${option}.`
                    } else {
                      return `${option}, `
                    }
                  }) }</p>

                  <h4>{ this.meatCount() + 'kg de carne' } <span role="img" aria-label="meat">üçñ</span></h4>
                  <h4>Costo total: ${ this.totalPrice().toLocaleString('es-ES') } <span role="img" aria-label="meat">üëà</span></h4>
                  <h4>Costo cada adulto: ${ this.eachPrice().toLocaleString('es-ES') } <span role="img" aria-label="meat">‚òù</span>Ô∏è</h4>
                </div>
              </div>
            }
          </div>
        </div>

        <Footer />
      </div>
    )
  }
}

export default Asado
