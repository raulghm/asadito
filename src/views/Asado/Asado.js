import React, { Component } from 'react'
import Footer from '../../components/Footer/Footer'
import './Asado.css'
import budgets from './budgets'

class Asado extends Component {
  constructor(props) {
    super(props)

    this.state = {
      title: '¬øCu√°nto debo comprar para el asado?',
      user: {
        men: 0,
        women: 0,
        children: 0,
      },
      breadPrice: 1200,
      carbonPrice: 1156,
      budgets,
      budgetIndex: 0,
      budgetSelected: null,
      sausageSelected: false,
      carbonSelected: false,
    }
  }

  handleMen = event => this.setState({
    user: {...this.state.user, men: parseInt(event.target.value, 10) }
  })

  handleWomen = event => this.setState({
    user: {...this.state.user, women: parseInt(event.target.value, 10) }
  })

  handleChildren = event => this.setState({
    user: {...this.state.user, children: parseInt(event.target.value, 10) }
  })

  handleSausage = () => this.setState({
    sausageSelected: !this.state.sausageSelected
  })

  handleCarbon = () => this.setState({
    carbonSelected: !this.state.carbonSelected
  })

  handleBudget = event => {
    const value = parseInt(event.target.value, 10)
    const budget = this.state.budgets.find(item => item.id === value)

    this.setState({
      budgetIndex: event.target.value,
      budgetSelected: budget
    })
  }

  peopleCount = () => {
    const computed = this.state.user.men +
      this.state.user.women +
      this.state.user.children

    return computed ? computed : 0
  }

  meatCount = () => {
    return this.round(
      (this.state.user.men * 0.35) +
      (this.state.user.women * 0.25) +
      (this.state.user.children * 0.2)
    )
  }

  sausageCount = () => {
    return this.round(
      (this.state.user.men * 0.1) +
      (this.state.user.women * 0.05) +
      (this.state.user.children * 0.05)
    )
  }

  breadCount = () => {
    const breadUnit = 0.08333

    return this.round(
      (this.state.user.men * breadUnit * 2) +
      (this.state.user.women * breadUnit * 2) +
      (this.state.user.children * breadUnit * 1)
    )
  }

  carbonCount = () => {
    return this.round(
      this.meatCount() +
      (this.state.sausageSelected ? this.sausageCount() : 0)
    )
  }

  round = (number, precision = 1) => {
    const factor = Math.pow(10, precision)
    const result = Math.round(number * factor) / factor
    return number ? result : 0
  }

  totalPrice = () => {
    const {
      budgetSelected,
      meatCount = this.meatCount(),
      sausageCount = this.sausageCount(),
      sausageSelected,
      carbonSelected,
      carbonPrice,
      breadCount = this.breadCount(),
      breadPrice,
      carbonCount = this.carbonCount(),
    } = this.state

    const meat = meatCount * budgetSelected.meatPrice

    const sausage = sausageSelected ?
                  (sausageCount * budgetSelected.sausagePrice) +
                  (breadCount * breadPrice) : 0

    const carbon = carbonSelected ?
                  carbonCount * carbonPrice : 0

    return this.round(
      meat +
      sausage +
      carbon
    , -1)
  }

  eachPrice = () => {
    return this.round(
      this.totalPrice() /
      (this.state.user.men + this.state.user.women)
    , -1)
  }

  render() {
    const {
      title,
      budgets,
      budgetSelected,
      peopleCount = this.peopleCount(),
      meatCount = this.meatCount(),
      sausageCount = this.sausageCount(),
      breadCount = this.breadCount(),
      carbonCount = this.carbonCount(),
      sausageSelected,
      carbonSelected,
    } = this.state

    return (
      <div className="Asado">
        <div className="Asado-wrapper">
          <div className="Asado-body">
            <h1>Calculadora de Asado <span role="img" aria-label="meat">üçñ</span></h1>

            <h2>{title}</h2>

            <div className="Asado-select">
              <h3>1- Indica la cantidad de comensales</h3>

              <div className="Asado-select-body Asado-select-body--column">
                <div className="Asado-select-item">
                  <h4>Hombres <span role="img" aria-label="men">üë®</span></h4>
                  <input
                    type="number"
                    pattern="[0-9]*"
                    min={0}
                    max={100}
                    placeholder="0"
                    onChange={this.handleMen.bind(this)}
                  />
                </div>

                <div className="Asado-select-item">
                  <h4>Mujeres <span role="img" aria-label="women">üë©</span></h4>
                  <input
                    type="number"
                    pattern="[0-9]*"
                    min={0}
                    max={100}
                    placeholder="0"
                    onChange={this.handleWomen.bind(this)}
                  />
                </div>

                <div className="Asado-select-item">
                  <h4>Ni√±os <span role="img" aria-label="children">üë∂</span></h4>
                  <input
                    type="number"
                    pattern="[0-9]*"
                    min={0}
                    max={100}
                    placeholder="0"
                    onChange={this.handleChildren.bind(this)}
                  />
                </div>
              </div>

              <h3>Total comensales: {peopleCount}</h3>
            </div>

            { peopleCount > 0 &&
              <div className="Asado-select">
                <h3>2- Selecciona tu presupuesto</h3>

                <div className="Asado-select-body Asado-select-body--column">
                  <div className="Asado-select-item">

                    <div className="Asado-input">
                      <select onChange={this.handleBudget.bind(this)}>
                        <option>Selecciona...</option>
                        { budgets.map(item => {
                          return <option
                            value={item.id}
                            key={item.id}
                          >{item.name}</option>
                        }) }
                      </select>
                    </div>

                    <div className="Asado-input Asado-input--checkbox">
                      <label>
                        <input
                          type="checkbox"
                          value={true}
                          onChange={this.handleSausage.bind(this)}
                        />
                        Incluir embutidos + pan
                      </label>
                    </div>

                    <div className="Asado-input Asado-input--checkbox">
                      <label>
                        <input
                          type="checkbox"
                          value={true}
                          onChange={this.handleCarbon.bind(this)}
                        />
                        Incluir carb√≥n
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            }

            { peopleCount > 0 && budgetSelected &&
              <div className="Asado-select">
                <h3>3- Resultados</h3>

                <div className="Asado-select-body">
                  Opciones de carnes para comprar:

                  <p>{ budgetSelected.options.map((option, i) => {
                    if (i === budgetSelected.options.length - 1) {
                      return `${option}.`
                    } else {
                      return `${option}, `
                    }
                  }) }</p>

                  <h4>
                    <ul>
                      <li>
                        { meatCount + 'kg de carne' }
                        <span role="img" aria-label="meat">üçñ</span>
                      </li>

                      { sausageSelected &&
                        <li>
                          { sausageCount + 'kg de embutidos' }
                        </li>
                      }

                      { sausageSelected &&
                        <li>
                          { breadCount + 'kg de pan' }
                        </li>
                      }

                      { carbonSelected &&
                        <li>
                          { carbonCount + 'kg de carb√≥n' }
                        </li>
                      }
                    </ul>
                  </h4>

                  <h4>
                    Costo total: ${ this.totalPrice().toLocaleString('es-ES') }
                    <span role="img" aria-label="meat">üëà</span>
                  </h4>

                  <h4>
                    Costo cada adulto: ${ this.eachPrice().toLocaleString('es-ES') }
                    <span role="img" aria-label="meat">‚òù</span>
                  </h4>
                </div>
              </div>
            }
          </div>
        </div>

        <Footer />
      </div>
    )
  }
}

export default Asado
