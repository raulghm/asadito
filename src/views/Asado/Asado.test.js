import React from 'react'
import Asado from './Asado'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

configure({ adapter: new Adapter() })

describe('<Asado /> shallow rendering', () => {
  it('should render Asado', () => {
    shallow(<Asado />)
  })
  it('on input change show budget and result', () => {
    const wrapper = shallow(<Asado />)

    const input = wrapper.find('.Asado-select--users input').first()
    input.simulate('change', { target: { value: '1' } })
    expect(wrapper.find('.Asado-select--budgets').length).toBe(1)

    const select = wrapper.find('.Asado-input select').first()
    select.simulate('change', { target: { value: 0 } })
    expect(wrapper.find('.Asado-select--results').length).toBe(1)
  })
})

describe('<Asado /> mount rendering', () => {
  it('should render Asado', () => {
    mount(<Asado />)
  })
  it('h2 contains correct text', () => {
    const wrapper = mount(<Asado />)
    expect(wrapper.find('h2').text()).toBe('¿Cuánto debo comprar para el asado?')
    wrapper.unmount()
  })
  it('matches the snapshot', () => {
    const tree = mount(<Asado />)
    expect(toJson(tree)).toMatchSnapshot()
    tree.unmount()
  })
})
