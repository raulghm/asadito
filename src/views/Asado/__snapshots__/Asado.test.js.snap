// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Asado /> mount rendering matches the snapshot 1`] = `
<Asado>
  <div
    className="Asado"
  >
    <div
      className="Asado-wrapper"
    >
      <div
        className="Asado-body"
      >
        <h1>
          Calculadora de Asado 
          <span
            aria-label="meat"
            role="img"
          >
            🍖
          </span>
        </h1>
        <h2>
          ¿Cuánto debo comprar para el asado?
        </h2>
        <div
          className="Asado-select Asado-select--users"
        >
          <h3>
            1- Indica la cantidad de comensales
          </h3>
          <div
            className="Asado-select-body Asado-select-body--column"
          >
            <div
              className="Asado-select-item"
            >
              <h4>
                <div>
                  Hombres
                </div>
                <span
                  aria-label="men"
                  role="img"
                >
                  👨
                </span>
              </h4>
              <input
                max={100}
                min={0}
                onChange={[Function]}
                pattern="[0-9]*"
                placeholder="0"
                type="number"
              />
            </div>
            <div
              className="Asado-select-item"
            >
              <h4>
                <div>
                  Mujeres
                </div>
                <span
                  aria-label="women"
                  role="img"
                >
                  👩
                </span>
              </h4>
              <input
                max={100}
                min={0}
                onChange={[Function]}
                pattern="[0-9]*"
                placeholder="0"
                type="number"
              />
            </div>
            <div
              className="Asado-select-item"
            >
              <h4>
                <div>
                  Niños
                </div>
                <span
                  aria-label="children"
                  role="img"
                >
                  👶
                </span>
              </h4>
              <input
                max={100}
                min={0}
                onChange={[Function]}
                pattern="[0-9]*"
                placeholder="0"
                type="number"
              />
            </div>
          </div>
          <h3>
            Total comensales: 
            0
          </h3>
        </div>
      </div>
    </div>
    <Footer>
      <div
        className="Footer"
      >
        raulghm 
        <span
          aria-label="ok_hand"
          role="img"
        >
          👌
        </span>
         
        <a
          href="https://github.com/raulghm/react-asado"
        >
          github.com/raulghm/react-asado
        </a>
      </div>
    </Footer>
  </div>
</Asado>
`;
